### PR TITLE
Log a warning for unused local variables and fix several data flow analysis issues

### LINKED ISSUE #33319: Emit warnings for local unused variables
**Description:**
See https://github.com/ballerina-platform/ballerina-spec/issues/356

### LINKED ISSUE #33485: Dataflow analysis doesn't run for the `expression` of a `member-access-lvexpr`
**Description:**
$title, where

> member-access-lvexpr := lvexpr [ expression ]

**Steps to reproduce:**
```ballerina
public function main() {
    int i;
    int[] m = [];

    m[i] = 1; // no error!

    int j = m[i]; // error: variable 'i' is not initialized

    int[] n;
    n[i] = 2; // error: variable 'n' is not initialized
}
```

**Affected Versions:**
slbeta3

### LINKED ISSUE #33495: Dataflow analysis doesn't run for the typedesc expression of an annotation access expression
**Description:**
$title. 

**Steps to reproduce:**

```ballerina
annotation v1 on type;

public function main()  {
    typedesc<any> t;
    _ = t.@v1;
}
```

Crashes at runtime instead.

```
[2021-10-28 19:35:19,180] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
        at io.ballerina.runtime.internal.TypeChecker.getAnnotValue(TypeChecker.java:578)
        at test.main(test.bal:5)
        at $_init.$lambda$main$(.)
        at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(Scheduler.java:610)
        at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:323)
        at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:291)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

**Affected Versions:**
slbeta3

### LINKED ISSUE #33508: Invalid `unused variable with inferred type including error` for variable used in infer-anonymous-function-expr
**Description:**
$title.

**Steps to reproduce:**
```ballerina
public function main() {
    var e1 = error(""); // no error

    function () returns error _ = function () returns error {
        return e1;
    };

    var e2 = error(""); // unused variable 'e2' with inferred type including error

    function () returns error _ = () => e2;
}
```

**Affected Versions:**
slbeta3

### LINKED ISSUE #33522: Dataflow analysis doesn't run for the type node of a local-no-init-var-decl-stmt
**Description:**
$title.

**Steps to reproduce:**
```ballerina
public function main() {
    final int j;
    record {| int i = j; |} a1; // no error!

    final int k;
    record {| int i = k; |} a2 = {}; // error as expected: variable 'k' is not initialized
}
```

**Affected Versions:**
slbeta4-rc1

### LINKED ISSUE #33524: Dataflow analysis doesn't run for match guards
**Description:**
$title.

**Steps to reproduce:**
```ballerina
public function main() {
    int[] & readonly arr = [];
    anydata[] anydataArr;

    match arr {
        [] if fn(...anydataArr) => { // no error
        }
    }
}

isolated function fn(any... x) returns boolean => true;
```

**Affected Versions:**
slbeta4-rc1

### LINKED ISSUE #33525: Dataflow analysis doesn't run for the function reference of a function call expression
**Description:**
$title.

**Steps to reproduce:**
```ballerina
public function main() {
    function () fn; 
    fn(); // no compilation error, crashes at runtime
}
```

**Affected Versions:**
slbeta4-rc1

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title.

Fixes #33319 - Emit warnings for local unused variables
Fixes #33525 - Data flow analysis doesn't run for the function reference of a function call expression
Fixes #33524 - Data flow analysis doesn't run for match guards
Fixes #33522 - Dataflow analysis doesn't run for the type node of a local-no-init-var-decl-stmt
Fixes #33508 - Invalid unused variable with inferred type including error for variable used in infer-anonymous-function-expr
Fixes #33495 - Dataflow analysis doesn't run for the typedesc expression of an annotation access expression
Fixes #33485 - Dataflow analysis doesn't run for the expression of a member-access-lvexpr

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: cli/ballerina-cli/src/test/resources/test-resources/validJava11Project/main.bal [hunk 1]
```diff
@@ -1,9 +1,9 @@
 import ballerina/jballerina.java;
 
 // Creates a file in the current directory.
-public function main(string filepath) {
+public function main(string filepath) returns error? {
    handle file = newFile(java:fromString(filepath));
-   boolean|error isSuccess = createNewFileInternal(file);
+   boolean _ = check createNewFileInternal(file);
```
Comment (MaryamZi) [inline cli/ballerina-cli/src/test/resources/test-resources/validJava11Project/main.bal:6]: @pramodya1994, hope this and other project API test changes are OK.

File: tests/jballerina-debugger-integration-test/src/test/resources/project-based-tests/debug-output-tests/main.bal [hunk 2]
```diff
@@ -16,5 +16,5 @@
 
 // The `main` function, which acts as the entry point.
 public function main() {
-    int x = 1/0;
+    int _ = 1/0;
```
Comment (MaryamZi) [inline tests/jballerina-debugger-integration-test/src/test/resources/project-based-tests/debug-output-tests/main.bal:19]: @NipunaRanasinghe, hope this change is OK.

Comment (NipunaRanasinghe) [reply tests/jballerina-debugger-integration-test/src/test/resources/project-based-tests/debug-output-tests/main.bal:19]: @MaryamZi test results of this test is independent from the variable names so this should be fine :)

**Review by NipunaRanasinghe — COMMENTED** [review]
Debugger test resources related changes LGTM

**Review by hasithaa — APPROVED** [review]

File: cli/ballerina-cli/src/test/resources/test-resources/validJava11Project/main.bal (hunk 1 above)
Comment (hasithaa) [reply cli/ballerina-cli/src/test/resources/test-resources/validJava11Project/main.bal:6]: Changes looks ok to me.

Comment (pramodya1994) [reply cli/ballerina-cli/src/test/resources/test-resources/validJava11Project/main.bal:6]: @MaryamZi Project API test resource changes look ok to me.

**Review by pramodya1994 — APPROVED** [review]
LGTM

**Pull request merged**